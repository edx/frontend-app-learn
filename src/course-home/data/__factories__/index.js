import './courseHomeMetadata.factory';
import './datesTabData.factory';
import './outlineTabData.factory';
