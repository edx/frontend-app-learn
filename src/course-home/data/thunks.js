import { logError } from '@edx/frontend-platform/logging';
import { camelCaseObject } from '@edx/frontend-platform';
import {
  executePostFromPostEvent,
  getCourseHomeCourseMetadata,
  getDatesTabData,
  getOutlineTabData,
  getProgressTabData,
  postCourseDeadlines,
  postDismissWelcomeMessage,
  postRequestCert,
} from './api';

import {
  addModel,
} from '../../generic/model-store';

import {
  fetchTabFailure,
  fetchTabRequest,
  fetchTabSuccess,
  setResetDatesToast,
} from './slice';

const eventTypes = {
  POST_EVENT: 'post_event',
};

export function fetchTab(courseId, tab, getTabData) {
  return async (dispatch) => {
    dispatch(fetchTabRequest({ courseId }));
    Promise.allSettled([
      getCourseHomeCourseMetadata(courseId),
      getTabData(courseId),
    ]).then(([courseHomeCourseMetadataResult, tabDataResult]) => {
      const fetchedCourseHomeCourseMetadata = courseHomeCourseMetadataResult.status === 'fulfilled';
      const fetchedTabData = tabDataResult.status === 'fulfilled';

      if (fetchedCourseHomeCourseMetadata) {
        dispatch(addModel({
          modelType: 'courses',
          model: {
            id: courseId,
            ...courseHomeCourseMetadataResult.value,
          },
        }));
      } else {
        logError(courseHomeCourseMetadataResult.reason);
      }

      if (fetchedTabData) {
        dispatch(addModel({
          modelType: tab,
          model: {
            id: courseId,
            ...tabDataResult.value,
          },
        }));
      } else {
        logError(tabDataResult.reason);
      }

      if (fetchedCourseHomeCourseMetadata && fetchedTabData) {
        dispatch(fetchTabSuccess({ courseId }));
      } else {
        dispatch(fetchTabFailure({ courseId }));
      }
    });
  };
}

export function fetchDatesTab(courseId) {
  return fetchTab(courseId, 'dates', getDatesTabData);
}

export function fetchProgressTab(courseId) {
  return fetchTab(courseId, 'progress', getProgressTabData);
}

export function fetchOutlineTab(courseId) {
  return fetchTab(courseId, 'outline', getOutlineTabData);
}

export function resetDeadlines(courseId, getTabData) {
  return async (dispatch) => {
    postCourseDeadlines(courseId).then(response => {
      const { data } = response;
      const {
        body,
        header,
      } = data;
      dispatch(getTabData(courseId));
      dispatch(setResetDatesToast({ body, header }));
    });
  };
}

export function dismissWelcomeMessage(courseId) {
  return async () => postDismissWelcomeMessage(courseId);
}

export function requestCert(courseId) {
  return async () => postRequestCert(courseId);
}

export function processEvent(eventData, getTabData) {
  return async (dispatch) => {
    const event = camelCaseObject(eventData);
    if (event.eventName === eventTypes.POST_EVENT) {
      executePostFromPostEvent(event.postData).then(response => {
        const { data } = response;
        const {
          body,
          header,
        } = data;
        dispatch(getTabData(event.postData.bodyParams.courseId));
        dispatch(setResetDatesToast({ body, header }));
      });
    }
  };
}
