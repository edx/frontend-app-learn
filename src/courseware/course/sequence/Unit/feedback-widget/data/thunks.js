import { logError } from '@edx/frontend-platform/logging';

import {
  createWholeCourseTranslationFeedbackFailure,
  createWholeCourseTranslationFeedbackRequest,
  createWholeCourseTranslationFeedbackSuccess,
  fetchWholeCourseTranslationFeedbackFailure,
  fetchWholeCourseTranslationFeedbackRequest,
  fetchWholeCourseTranslationFeedbackSuccess,
} from './slice';
import { createTranslationFeedback, getTranslationFeedback } from './api';
import { updateModel } from '../../../../../../generic/model-store';

export function fetchWholeCourseTranslationFeedback(
  courseId,
  translationLanguage,
  unitId,
  userId,
) {
  return async (dispatch) => {
    dispatch(fetchWholeCourseTranslationFeedbackRequest({
      courseId,
    }));
    try {
      const existingFeedback = await getTranslationFeedback(
        courseId,
        translationLanguage,
        unitId,
        userId,
      );
      dispatch(updateModel({
        modelType: 'coursewareMeta',
        model: {
          id: courseId,
          translationFeedback: existingFeedback ? {
            translationLanguage,
            unitId,
            userId,
            feedbackValue: existingFeedback?.feedbackValue,
          } : null,
        },
      }));
      dispatch(fetchWholeCourseTranslationFeedbackSuccess({
        courseId,
      }));
    } catch (error) {
      logError(error);
      dispatch(fetchWholeCourseTranslationFeedbackFailure({
        courseId,
      }));
    }
  };
}

export function createWholeCourseTranslationFeedback(
  courseId,
  feedbackValue,
  translationLanguage,
  unitId,
  userId,
) {
  return async (dispatch) => {
    dispatch(createWholeCourseTranslationFeedbackRequest({
      courseId,
    }));
    try {
      const newFeedback = await createTranslationFeedback(
        courseId,
        feedbackValue,
        translationLanguage,
        unitId,
        userId,
      );
      dispatch(updateModel({
        modelType: 'coursewareMeta',
        model: {
          id: courseId,
          translationFeedback: {
            translationLanguage,
            unitId,
            userId,
            feedbackValue: newFeedback.feedbackValue,
          },
        },
      }));
      dispatch(createWholeCourseTranslationFeedbackSuccess({
        courseId,
      }));
    } catch (error) {
      logError(error);
      dispatch(createWholeCourseTranslationFeedbackFailure({
        courseId,
      }));
    }
  };
}
