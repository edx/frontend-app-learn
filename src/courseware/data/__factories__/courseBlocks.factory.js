import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

import './block.factory';

Factory.define('courseBlocks')
  .option('courseId', 'course-v1:edX+DemoX+Demo_Course')
  .option('unit', ['courseId'], courseId => Factory.build(
    'block',
    { type: 'vertical' },
    { courseId },
  ))
  .option('sequence', ['courseId', 'unit'], (courseId, child) => Factory.build(
    'block',
    { type: 'sequential', children: [child.id] },
    { courseId },
  ))
  .option('section', ['courseId', 'sequence'], (courseId, child) => Factory.build(
    'block',
    { type: 'chapter', children: [child.id] },
    { courseId },
  ))
  .option('course', ['courseId', 'section'], (courseId, child) => Factory.build(
    'block',
    { type: 'course', children: [child.id] },
    { courseId },
  ))
  .attr(
    'blocks',
    ['course', 'section', 'sequence', 'unit'],
    (course, section, sequence, unit) => ({
      [course.id]: course,
      [section.id]: section,
      [sequence.id]: sequence,
      [unit.id]: unit,
    }),
  )
  .attr('root', ['course'], course => course.id);
