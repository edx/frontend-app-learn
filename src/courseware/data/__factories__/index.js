import './courseMetadata.factory';
import './sequenceMetadata.factory';
import './courseRecommendations.factory';
