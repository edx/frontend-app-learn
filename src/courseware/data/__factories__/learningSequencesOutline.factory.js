import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

Factory.define('learningSequencesOutline')
  .attrs({
    title: 'Demo Course',
    course_id: 'course-v1:edX+DemoX+Demo_Course',
    outline: {
      sections: [],
      sequences: {
      },
    },
  });

export function buildEmptyOutline(courseId) {
  return Factory.build(
    'learningSequencesOutline',
    {
      title: 'Demo Course',
      course_id: courseId,
      outline: {
        sections: [],
        sequences: {
        },
      },
    },
    { courseId },
  );
}

// Given courseBlocks (output from buildSimpleCourseBlocks), create a matching
// Learning Sequences API outline (what the REST API would return to us).
export function buildOutlineFromBlocks(courseBlocks) {
  const sections = {};
  const sequences = {};
  let courseBlock = null;

  Object.values(courseBlocks.blocks).forEach(block => {
    if (block.type === 'course') {
      courseBlock = block;
    } else if (block.type === 'chapter') {
      sections[block.id] = {
        id: block.id,
        title: block.title,
        start: null,
        sequence_ids: [...block.children],
      };
    } else if (block.type === 'sequential') {
      sequences[block.id] = {
        id: block.id,
        title: block.title,
        accessible: true,
        start: null,
      };
    }
  });

  const outline = Factory.build(
    'learningSequencesOutline',
    {
      course_key: courseBlocks.courseId,
      title: courseBlocks.title,
      outline: {
        sections: courseBlock.children.map(sectionId => sections[sectionId]),
        sequences,
      },
    },
    {},
  );

  return outline;
}
