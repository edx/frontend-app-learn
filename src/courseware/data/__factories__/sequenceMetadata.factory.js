import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

import './block.factory';

Factory.define('sequenceMetadata')
  .option('courseId', (courseId) => {
    if (courseId) {
      return courseId;
    }
    throw new Error('courseId must be specified for sequenceMetadata factory.');
  })
  // An array of units
  .option('unitBlocks', ['courseId'], courseId => ([
    Factory.build(
      'block',
      { type: 'vertical' },
      { courseId },
    ),
  ]))
  .option(
    'sequenceBlock', ['courseId', 'unitBlocks'], (courseId, unitBlocks) => (
      Factory.build(
        'block',
        { type: 'sequential', children: unitBlocks.map(unitBlock => unitBlock.id) },
        { courseId },
      )
    ),
  )
  .attr('element_id', ['sequenceBlock'], sequenceBlock => sequenceBlock.block_id)
  .attr('item_id', ['sequenceBlock'], sequenceBlock => sequenceBlock.id)
  .attr('display_name', ['sequenceBlock'], sequenceBlock => sequenceBlock.display_name)
  .attr('ajax_url', ['sequenceBlock'], sequenceBlock => `${sequenceBlock.student_view_url}/handler/xmodule_handler}`)
  .attr('gated_content', ['sequenceBlock'], sequenceBlock => ({
    gated: false,
    prereq_url: null,
    prereq_id: null,
    prereq_section_name: null,
    gated_section_name: sequenceBlock.display_name,
  }))
  .attr('items', ['unitBlocks', 'sequenceBlock'], (unitBlocks, sequenceBlock) => unitBlocks.map(
    unitBlock => ({
      href: '',
      graded: unitBlock.graded,
      id: unitBlock.id,
      bookmarked: false,
      path: `Chapter Display Name > ${sequenceBlock.display_name} > ${unitBlock.display_name}`,
      type: 'other',
      complete: null,
      content: '',
      page_title: unitBlock.display_name,
    }),
  ))
  .attrs({
    exclude_units: true,
    position: null,
    next_url: null,
    tag: 'sequential',
    save_position: true,
    prev_url: null,
    is_time_limited: false,
    show_completion: true,
    banner_text: null,
  });
