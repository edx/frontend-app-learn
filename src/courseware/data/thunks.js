import { logError } from '@edx/frontend-platform/logging';
import {
  getBlockCompletion,
  postSequencePosition,
} from './api';
import {
  updateModel,
} from '../../generic/model-store';

export function checkBlockCompletion(courseId, sequenceId, unitId) {
  return async (dispatch, getState) => {
    const { models } = getState();
    if (models.units[unitId].complete) {
      return; // do nothing. Things don't get uncompleted after they are completed.
    }

    try {
      const isComplete = await getBlockCompletion(courseId, sequenceId, unitId);
      dispatch(updateModel({
        modelType: 'units',
        model: {
          id: unitId,
          complete: isComplete,
        },
      }));
    } catch (error) {
      logError(error);
    }
  };
}

export function saveSequencePosition(courseId, sequenceId, position) {
  return async (dispatch, getState) => {
    const { models } = getState();
    const initialPosition = models.sequences[sequenceId].position;
    // Optimistically update the position.
    dispatch(updateModel({
      modelType: 'sequences',
      model: {
        id: sequenceId,
        position,
      },
    }));
    try {
      await postSequencePosition(courseId, sequenceId, position);
      // Update again under the assumption that the above call succeeded, since it doesn't return a
      // meaningful response.
      dispatch(updateModel({
        modelType: 'sequences',
        model: {
          id: sequenceId,
          position,
        },
      }));
    } catch (error) {
      logError(error);
      dispatch(updateModel({
        modelType: 'sequences',
        model: {
          id: sequenceId,
          position: initialPosition,
        },
      }));
    }
  };
}
