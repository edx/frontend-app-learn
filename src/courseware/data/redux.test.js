import { Factory } from 'rosie';
import MockAdapter from 'axios-mock-adapter';

import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';

import * as thunks from './thunks';

import executeThunk from '../../utils';

import buildSimpleCourseBlocks from './__factories__/courseBlocks.factory';
import './__factories__';
import initializeMockApp from '../../setupTest';
import initializeStore from '../../store';

const { loggingService } = initializeMockApp();

const axiosMock = new MockAdapter(getAuthenticatedHttpClient());

describe('Data layer integration tests', () => {
  const courseBaseUrl = `${getConfig().LMS_BASE_URL}/api/courseware/course`;
  const courseBlocksUrlRegExp = new RegExp(`${getConfig().LMS_BASE_URL}/api/courses/v2/blocks/*`);
  const sequenceBaseUrl = `${getConfig().LMS_BASE_URL}/api/courseware/sequence`;

  // building minimum set of api responses to test all thunks
  const courseMetadata = Factory.build('courseMetadata');
  const courseId = courseMetadata.id;
  const { courseBlocks, unitBlock, sequenceBlock } = buildSimpleCourseBlocks(courseId);
  const sequenceMetadata = Factory.build(
    'sequenceMetadata',
    { courseId },
    { unitBlock, sequenceBlock },
  );

  const courseUrl = `${courseBaseUrl}/${courseId}`;
  const sequenceUrl = `${sequenceBaseUrl}/${sequenceMetadata.item_id}`;
  const sequenceId = sequenceBlock.id;
  const unitId = unitBlock.id;

  let store;

  beforeEach(() => {
    axiosMock.reset();
    loggingService.logError.mockReset();

    store = initializeStore();
  });

  describe('Test fetchCourse', () => {
    it('Should fail to fetch course and blocks if request error happens', async () => {
      axiosMock.onGet(courseUrl).networkError();
      axiosMock.onGet(courseBlocksUrlRegExp).networkError();

      await executeThunk(thunks.fetchCourse(courseId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseware).toEqual(expect.objectContaining({
        courseId,
        courseStatus: 'failed',
      }));
    });

    it('Should fetch, normalize, and save metadata, but with denied status', async () => {
      const forbiddenCourseMetadata = Factory.build('courseMetadata', {
        can_load_courseware: {
          has_access: false,
        },
      });
      const forbiddenCourseBlocks = Factory.build('courseBlocks', { courseId: forbiddenCourseMetadata.id });

      const forbiddenCourseUrl = `${courseBaseUrl}/${forbiddenCourseMetadata.id}`;

      axiosMock.onGet(forbiddenCourseUrl).reply(200, forbiddenCourseMetadata);
      axiosMock.onGet(courseBlocksUrlRegExp).reply(200, forbiddenCourseBlocks);

      await executeThunk(thunks.fetchCourse(forbiddenCourseMetadata.id), store.dispatch);

      const state = store.getState();

      expect(state.courseware.courseStatus).toEqual('denied');

      // check that at least one key camel cased, thus course data normalized
      expect(state.models.courses[forbiddenCourseMetadata.id].canLoadCourseware).not.toBeUndefined();
    });

    it('Should fetch, normalize, and save metadata', async () => {
      axiosMock.onGet(courseUrl).reply(200, courseMetadata);
      axiosMock.onGet(courseBlocksUrlRegExp).reply(200, courseBlocks);

      await executeThunk(thunks.fetchCourse(courseId), store.dispatch);

      const state = store.getState();

      expect(state.courseware.courseStatus).toEqual('loaded');
      expect(state.courseware.courseId).toEqual(courseId);
      expect(state.courseware.sequenceStatus).toEqual('loading');
      expect(state.courseware.sequenceId).toEqual(null);

      // check that at least one key camel cased, thus course data normalized
      expect(state.models.courses[courseId].canLoadCourseware).not.toBeUndefined();
    });
  });

  describe('Test fetchSequence', () => {
    it('Should result in fetch failure if error occurs', async () => {
      axiosMock.onGet(sequenceUrl).networkError();

      await executeThunk(thunks.fetchSequence(sequenceId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseware.sequenceStatus).toEqual('failed');
    });

    it('Should fetch and normalize metadata, and then update existing models with sequence metadata', async () => {
      axiosMock.onGet(courseUrl).reply(200, courseMetadata);
      axiosMock.onGet(courseBlocksUrlRegExp).reply(200, courseBlocks);
      axiosMock.onGet(sequenceUrl).reply(200, sequenceMetadata);

      // setting course with blocks before sequence to check that blocks receive
      // additional information after fetchSequence call.
      await executeThunk(thunks.fetchCourse(courseId), store.dispatch);

      // ensure that initial state has no additional sequence info
      let state = store.getState();
      expect(state.models.sequences).toEqual({
        [sequenceBlock.id]: expect.not.objectContaining({
          gatedContent: expect.any(Object),
          activeUnitIndex: expect.any(Number),
        }),
      });
      expect(state.models.units).toEqual({
        [unitBlock.id]: expect.not.objectContaining({
          complete: null,
          bookmarked: expect.any(Boolean),
        }),
      });

      // Update our state variable again.
      state = store.getState();

      expect(state.courseware.courseStatus).toEqual('loaded');
      expect(state.courseware.courseId).toEqual(courseId);
      expect(state.courseware.sequenceStatus).toEqual('loading');
      expect(state.courseware.sequenceId).toEqual(null);

      await executeThunk(thunks.fetchSequence(sequenceBlock.id), store.dispatch);

      // Update our state variable again.
      state = store.getState();

      // ensure that additional information appeared in store
      expect(state.models.sequences).toEqual({
        [sequenceBlock.id]: expect.objectContaining({
          gatedContent: expect.any(Object),
          activeUnitIndex: expect.any(Number),
        }),
      });
      expect(state.models.units).toEqual({
        [unitBlock.id]: expect.objectContaining({
          complete: null,
          bookmarked: expect.any(Boolean),
        }),
      });

      expect(state.courseware.courseStatus).toEqual('loaded');
      expect(state.courseware.courseId).toEqual(courseId);
      expect(state.courseware.sequenceStatus).toEqual('loaded');
      expect(state.courseware.sequenceId).toEqual(sequenceId);
    });
  });

  describe('Thunks that require fetched sequences', () => {
    beforeEach(async () => {
      // thunks tested in this block rely on fact, that store already has
      // some info about sequence
      axiosMock.onGet(sequenceUrl).reply(200, sequenceMetadata);
      await executeThunk(thunks.fetchSequence(sequenceMetadata.item_id), store.dispatch);
    });

    describe('Test checkBlockCompletion', () => {
      const getCompletionURL = `${getConfig().LMS_BASE_URL}/courses/${courseId}/xblock/${sequenceId}/handler/xmodule_handler/get_completion`;

      it('Should fail to check completion and log error', async () => {
        axiosMock.onPost(getCompletionURL).networkError();

        await executeThunk(
          thunks.checkBlockCompletion(courseId, sequenceId, unitId),
          store.dispatch,
          store.getState,
        );

        expect(loggingService.logError).toHaveBeenCalled();
        expect(axiosMock.history.post[0].url).toEqual(getCompletionURL);
      });

      it('Should update complete field of unit model', async () => {
        axiosMock.onPost(getCompletionURL).reply(201, { complete: true });

        await executeThunk(
          thunks.checkBlockCompletion(courseId, sequenceId, unitId),
          store.dispatch,
          store.getState,
        );

        expect(store.getState().models.units[unitId].complete).toBeTruthy();
      });
    });

    describe('Test saveSequencePosition', () => {
      const gotoPositionURL = `${getConfig().LMS_BASE_URL}/courses/${courseId}/xblock/${sequenceId}/handler/xmodule_handler/goto_position`;

      it('Should change and revert sequence model position in case of error', async () => {
        axiosMock.onPost(gotoPositionURL).networkError();

        const oldPosition = store.getState().models.sequences[sequenceId].position;
        const newPosition = 123;

        await executeThunk(
          thunks.saveSequencePosition(courseId, sequenceId, newPosition),
          store.dispatch,
          store.getState,
        );

        expect(loggingService.logError).toHaveBeenCalled();
        expect(axiosMock.history.post[0].url).toEqual(gotoPositionURL);
        expect(store.getState().models.sequences[sequenceId].position).toEqual(oldPosition);
      });

      it('Should update sequence model position', async () => {
        axiosMock.onPost(gotoPositionURL).reply(201, {});

        const newPosition = 123;

        await executeThunk(
          thunks.saveSequencePosition(courseId, sequenceId, newPosition),
          store.dispatch,
          store.getState,
        );

        expect(axiosMock.history.post[0].url).toEqual(gotoPositionURL);
        expect(store.getState().models.sequences[sequenceId].position).toEqual(newPosition);
      });
    });
  });
});
