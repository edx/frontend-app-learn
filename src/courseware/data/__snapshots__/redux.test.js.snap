// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data layer integration tests Test fetchCourse Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseware": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course_1",
    "courseStatus": "loaded",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courses": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "canLoadCourseware": Object {
          "additionalContextUserMessage": null,
          "developerMessage": null,
          "errorCode": null,
          "hasAccess": true,
          "userFragment": null,
          "userMessage": null,
        },
        "canShowUpgradeSock": false,
        "celebrations": undefined,
        "contentTypeGatingEnabled": false,
        "courseExpiredMessage": null,
        "end": null,
        "enrollmentEnd": null,
        "enrollmentMode": null,
        "enrollmentStart": null,
        "id": "course-v1:edX+DemoX+Demo_Course_1",
        "isEnrolled": null,
        "isStaff": false,
        "marketingUrl": null,
        "notes": Object {
          "enabled": false,
          "visible": true,
        },
        "number": "DemoX",
        "offerHtml": null,
        "org": "edX",
        "sectionIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        ],
        "showCalculator": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "priority": 0,
            "slug": "courseware",
            "title": "Course",
            "type": "courseware",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course/",
          },
          Object {
            "priority": 1,
            "slug": "discussion",
            "title": "Discussion",
            "type": "discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/discussion/forum/",
          },
          Object {
            "priority": 2,
            "slug": "wiki",
            "title": "Wiki",
            "type": "wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course_wiki",
          },
          Object {
            "priority": 3,
            "slug": "progress",
            "title": "Progress",
            "type": "progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/progress",
          },
          Object {
            "priority": 4,
            "slug": "instructor",
            "title": "Instructor",
            "type": "instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/instructor",
          },
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd4",
        "verifiedMode": Object {
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "sections": Object {
      "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3": Object {
        "courseId": "course-v1:edX+DemoX+Demo_Course_1",
        "id": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        "sequenceIds": Array [
          "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd3",
      },
    },
    "sequences": Object {
      "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2": Object {
        "id": "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/jump_to/block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "sectionId": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd2",
        "unitIds": Array [
          "block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        ],
      },
    },
    "units": Object {
      "block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1": Object {
        "graded": false,
        "id": "block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/jump_to/block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        "sequenceId": "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd1",
      },
    },
  },
}
`;

exports[`Data layer integration tests Test fetchSequence Should fetch and normalize metadata, and then update existing models with sequence metadata 1`] = `
Object {
  "courseware": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course_1",
    "courseStatus": "loaded",
    "sequenceId": "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
    "sequenceStatus": "loaded",
  },
  "models": Object {
    "courses": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "canLoadCourseware": Object {
          "additionalContextUserMessage": null,
          "developerMessage": null,
          "errorCode": null,
          "hasAccess": true,
          "userFragment": null,
          "userMessage": null,
        },
        "canShowUpgradeSock": false,
        "celebrations": undefined,
        "contentTypeGatingEnabled": false,
        "courseExpiredMessage": null,
        "end": null,
        "enrollmentEnd": null,
        "enrollmentMode": null,
        "enrollmentStart": null,
        "id": "course-v1:edX+DemoX+Demo_Course_1",
        "isEnrolled": null,
        "isStaff": false,
        "marketingUrl": null,
        "notes": Object {
          "enabled": false,
          "visible": true,
        },
        "number": "DemoX",
        "offerHtml": null,
        "org": "edX",
        "sectionIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        ],
        "showCalculator": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "priority": 0,
            "slug": "courseware",
            "title": "Course",
            "type": "courseware",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course/",
          },
          Object {
            "priority": 1,
            "slug": "discussion",
            "title": "Discussion",
            "type": "discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/discussion/forum/",
          },
          Object {
            "priority": 2,
            "slug": "wiki",
            "title": "Wiki",
            "type": "wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course_wiki",
          },
          Object {
            "priority": 3,
            "slug": "progress",
            "title": "Progress",
            "type": "progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/progress",
          },
          Object {
            "priority": 4,
            "slug": "instructor",
            "title": "Instructor",
            "type": "instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/instructor",
          },
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd4",
        "verifiedMode": Object {
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "sections": Object {
      "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3": Object {
        "courseId": "course-v1:edX+DemoX+Demo_Course_1",
        "id": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        "sequenceIds": Array [
          "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd3",
      },
    },
    "sequences": Object {
      "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2": Object {
        "activeUnitIndex": 0,
        "bannerText": null,
        "gatedContent": Object {
          "gated": false,
          "gatedSectionName": "bcdabcdabcdabcdabcdabcdabcdabcd2",
          "prereqId": null,
          "prereqSectionName": null,
          "prereqUrl": null,
        },
        "id": "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "isTimeLimited": false,
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/jump_to/block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "position": 1,
        "saveUnitPosition": true,
        "sectionId": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        "showCompletion": true,
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd2",
        "unitIds": Array [
          "block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        ],
      },
    },
    "units": Object {
      "block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1": Object {
        "bookmarked": false,
        "complete": null,
        "contentType": "other",
        "graded": false,
        "id": "block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/jump_to/block-v1:edX+DemoX+Demo_Course_1+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        "sequenceId": "block-v1:edX+DemoX+Demo_Course_1+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd1",
      },
    },
  },
}
`;
